### PR TITLE
Add test for allocate nodes, minor test fixes

### PR DESCRIPTION
**Pull request by felipe-pizzorno**

### CODE REVIEW

File: crates/core/src/octree/build/mod.rs [hunk 1]
```diff
@@ -181,13 +183,15 @@ impl Octree {
                 let geometry_first_node_in_level =
                     geometry_level_start_indices[octree_level as usize];
 
-                self.builder.allocate_nodes_pass.run(
-                    &self.geometry_data.voxel_data, // TODO: Pass in the actual number of nodes
-                    &self.textures,
+                let allocate_nodes_input = AllocateNodesInput {
+                    voxel_data: self.border_data.voxel_data.clone(),
```
Comment (felipe-pizzorno) [inline crates/core/src/octree/build/mod.rs:187]: Changed geometry_data to border_data, although it makes sense it can be a breaking change (will test before merging)

File: crates/core/src/config.rs [hunk 2]
```diff
@@ -30,7 +30,12 @@ impl Config {
     /// Initializes the config
     /// Must be called before any call to `instance`
     pub unsafe fn initialize(mut config: Self) {
-        if INSTANCE.get().is_some() { panic!("Can only initialize core config once"); }
+        Self::initialize_test_sensitive(config, false);
+    }
+
+    pub unsafe fn initialize_test_sensitive(mut config: Self, is_test: bool) {
+        // We may initialize many times in test mode
+        if INSTANCE.get().is_some() && !is_test { panic!("Can only initialize core config once"); }
```
Comment (franciscoaguirre) [inline crates/core/src/config.rs:38]: You could do:

```rust
#[cfg(not(test))]
pub unsafe fn initialize(mut config: Self) { ... }

#[cfg(test)]
pub unsafe fn initialize(mut config: Self) { ... }
```

This would make the test initialize different

File: crates/core/src/octree/build/stages/allocate_nodes.rs [hunk 3]
```diff
@@ -57,3 +63,157 @@ impl AllocateNodesPass {
...
+
+        let voxel_dimension_exponent = 4;
+
+        unsafe {
+            let test_data = test_cases();
+
+            // Initialize everything
+            Config::initialize_test_sensitive(Config::new(voxel_dimension_exponent), true);
+            let allocate_nodes_pass = AllocateNodesPass::init();
+            let mut allocations: Vec<u32> = vec!();
```
Comment (franciscoaguirre) [inline crates/core/src/octree/build/stages/allocate_nodes.rs]:
```suggestion
            let mut allocations: Vec<u32> = vec![];
```

I've just never seen it with parenthesis before 😅, I know it works

File: crates/core/src/octree/build/stages/allocate_nodes.rs [hunk 4]
```diff
@@ -57,3 +63,157 @@ impl AllocateNodesPass {
...
+                let expected_allocations: Vec<u32> = (input.first_free_node as u32..input.first_free_node as u32 + allocations.len() as u32 ).collect();
+                allocations.sort();
+                assert_eq!(*allocations, *expected_allocations);
+                allocations.clear();
+                println!("Passed");
+            }
+        }
+    }
+
+    unsafe fn test_cases() -> std::vec::Vec<TestCase> {
```
Comment (franciscoaguirre) [inline crates/core/src/octree/build/stages/allocate_nodes.rs:125]: I'd love to test the case where we add border data to the mix
We should also look into what edge cases we can test, such as number of threads

File: crates/core/src/octree/build/stages/flag_nodes.rs [hunk 5]
```diff
@@ -72,19 +72,13 @@ mod tests {
     fn flag_nodes_works() {
         let (_glfw, _window) = test_utils::init_opengl_context();
 
-        // To go from the crate root to the workspace root
-        let mut path = PathBuf::from(env::current_dir().unwrap());
-        path.pop();
-        path.pop();
-        env::set_current_dir(path).unwrap();
-
```
Comment (franciscoaguirre) [inline crates/core/src/octree/build/stages/flag_nodes.rs:80]: A hack we could do is save the cwd, do this at the start and then go back to the previous cwd at the end of the test. I'll try to do that. Created https://github.com/franciscoaguirre/voxel-cone-tracing/issues/99 for it

File: crates/core/src/config.rs (hunk 2 above)
Comment (felipe-pizzorno) [reply crates/core/src/config.rs:38]: I still wanted to test that the config can't be initialized twice, didn't know how to do that with the `#[cfg(test)]` approach

File: crates/core/src/octree/build/stages/allocate_nodes.rs (hunk 4 above)
Comment (felipe-pizzorno) [reply crates/core/src/octree/build/stages/allocate_nodes.rs:125]: That is probably more in the side of integration tests right? For border data

Comment (franciscoaguirre) [reply crates/core/src/octree/build/stages/allocate_nodes.rs:125]: Probably

**Review by franciscoaguirre — APPROVED** [review]
